bpf/bpf_overlay.c

```diff
@@ -487,8 +487,8 @@ int tail_handle_arp(struct __ctx_buff *ctx)
 		return send_drop_notify_error(ctx, 0, ret, CTX_ACT_DROP, METRIC_EGRESS);
 	if (info->tunnel_endpoint) {
 		ret = __encap_and_redirect_with_nodeid(ctx, 0, info->tunnel_endpoint,
-						       LOCAL_NODE_ID, WORLD_ID,
-						       WORLD_ID, &trace);
+						       LOCAL_NODE_ID, WORLD_IPV4_ID,
+						       WORLD_IPV4_ID, &trace);
```

Comment:
Hmm, setting the VNI field in the packet to `WORLD_IPV4_ID` could potentially cause confusion for other agents in the cluster when transmitting these packets, until those peer agents have also upgraded to a version that understands `WORLD_IPV4_ID`. How is this code exercised?

Reply:
Probably what this should do is transmit WORLD_ID and then the receive side should decide to treat "WORLD_ID" as "WORLD_IPV4_ID" iff dualstack is enabled at the destination. That would ensure that this doesn't cause connectivity issues if the peer node doesn't understand WORLD_IPV4_ID upon upgrade.

Reply:
(discovered while freshly looking at the multiprotocol SECLABEL handling in the PR, but I didn't search around for whether there might be any other similar cases)

Reply:
So it will only be for dual-stack. It is an upgrade issue that I hadn't considered, but I don't see any way around it.

Reply:
Hmm, digging back through the git history, it looks like this encap identity here is specifically for the VTEP support. Using any VNI other than 2 may just break that feature. I think the right solution here is to keep the second to last parameter as `WORLD_ID`. The intermediate parameter can stay as `WORLD_IPV4_ID`, since that is used for local trace notifications only.

Even if we wanted to transition this to transmitting a different identity, I think that in general we should be introducing that gradually to minimize cluster disruption during the upgrades. But in this particular case with VTEP I don't even see a path for that.

Reply:
[It looks like the `tunnel_id` is also set to the SECLABEL when VTEP is not enabled](https://github.com/cilium/cilium/blob/b3a19b54269435248419d2cbe6ff42afd9b479b7/bpf/lib/overloadable_skb.h#L202). Woulnd't that also be a problem?

Reply:
Hmm yeah I didn't step deep enough. 

My gut feeling here is that the right approach for the backportable fix is to set both to WORLD_ID. Even if we wanted to start transmitting the newer identity, that change should only be introduced in v1.15 after we backported this to at least v1.14.

Then it's a question of exactly which paths care about handling ARP from a tunnel peer. I think the reason it's important in a VTEP case is that the VTEP doesn't have another out-of-bound mechanism to learn Pod IP locations. However I'm not sure whether Cilium ever sends ARP like this over the tunnels. If this is purely for VTEP then in this case I think it's both (a) fine to leave these as WORLD_ID and (b) there's no further changes necessary, it's the correct fix.

On the other hand if there is a non-VTEP case where this path is expected to do ARP response then the peer's handling of the security identity should be further inspected.

Reply:
I'll try this out, but I got tunneling errors in CI before I made the SECLABEL changes.

Reply:
Those tunneling errors could potentially be the first sign of upgrade compatibility impact, because new identities were being transmitted and the corresponding receive paths were not handling them. Hard to say without seeing them directly though.